cache/fast_lru_cache.h

```diff
@@ -22,31 +23,106 @@
+//    happen if the element is V, since references to an element can only be
+//    created when it is visible.
+
+// Internally, the cache uses an open-addressed hash table to index the handles.
+// We use tombstone counters to keep track of displacements.
+// Because of the tombstones and the two possible visibility states of an
+// element, the table slots can be in 4 different states:
+// 1. Visible element (IS_ELEMENT set and IS_VISIBLE set): The slot contains a
+//    key-value element.
+// 2. Ghost element (IS_ELEMENT set and IS_VISIBLE unset): The slot contains an
+//    element that has been removed, but it's still referenced. It's invisible
+//    to lookups.
+// 3. Tombstone (IS_ELEMENT unset and displacements > 0): The slot contains a
+//    tombstone.
+// 4. Empty (IS_ELEMENT unset and displacements == 0): The slot is unused.
+//    A slot that is an element can further have IS_VISIBLE set or not.
+// When a ghost is removed from the table, it can either transition to being a
+// tombstone or an empty slot, depending on the number of displacements of the
+// slot. In any case, the slot becomes available. When a handle is inserted
+// into that slot, it becomes a visible element again.
+
+constexpr uint8_t kCacheKeySize =
+    static_cast<uint8_t>(sizeof(ROCKSDB_NAMESPACE::CacheKey));
+
+// The load factor should be at least 50% and far from 100%.
+// Load factor p implies that in a full table (i.e., a fraction
+// p of all slots, without counting tombstones, are occupied by
+// elements) the probability that a random probe hits an empty
+// slot is p, and thus 1/p probes are required on average.
+// For p = 70%, between 1 and 2 probes are needed on average.
+constexpr double kLoadFactor = 0.7;
+
+constexpr uint32_t kProbingSeed = 0x7a2bb9d5; // Arbitrary seeds.
+constexpr uint32_t kLargePrime = 433494437; // Arbitrary large prime number.
+
 // An experimental (under development!) alternative to LRUCache
 
 struct LRUHandle {
   void* value;
   Cache::DeleterFn deleter;
-  LRUHandle* next_hash;
   LRUHandle* next;
   LRUHandle* prev;
   size_t total_charge;  // TODO(opt): Only allow uint32_t?
-  size_t key_length;
   // The hash of key(). Used for fast sharding and comparisons.
   uint32_t hash;
   // The number of external refs to this entry. The cache itself is not counted.
   uint32_t refs;
 
   enum Flags : uint8_t {
-    // Whether this entry is referenced by the hash table.
-    IN_CACHE = (1 << 0),
+    // Whether the handle is visible to Lookups.
+    IS_VISIBLE = (1 << 0),
+    // Whether the slot is in use by an element.
+    IS_ELEMENT = (1 << 1),
   };
   uint8_t flags;
 
-  // Beginning of the key (MUST BE THE LAST FIELD IN THIS STRUCT!)
-  char key_data[1];
+  // The number of elements that hash to this slot or a lower one,
+  // but wind up in a higher slot.
+  uint32_t displacements;
+
+  char key_data[kCacheKeySize];
+
+  LRUHandle() { memset(this, 0, sizeof(LRUHandle)); }
```

Comment:
This is not really safe / kosher if LRUHandle should ever hold non-POD data. I recommend doing `= 0` or `= nullptr` or `= {}` on all the field definitions.